### PR TITLE
LL-7852 Improve BTC-JS sync perf

### PR DESCRIPTION
**Pull request by ghost**
## Context (issues, jira)

https://ledgerhq.atlassian.net/browse/LL-7852

### :white_check_mark: During sync replace use of `Array.includes` (time complexity o(n)) with `Set.has` (time complexity o(1)).

Sync with LL CLI of a large BTC xpub (~12000 ops) :

- libcore : 1m42,047s
- JS without this PR: 2m16,161s
- JS with this PR: 1m45,314s

Perf diff on the mapTxToOperations function
- before:
Screenshot from 2021-12-23 14-42-43

- after :
Screenshot from 2021-12-23 14-42-32

### :white_check_mark: Rework some parts in the wallet-btc implementation

- [ ] Rework complexity of implementation where relevant (e.g use of lodash)
- [ ] Rework use of heavy cryptographic functions (e.g from bitcoinjs-lib)

## Expectations

- [x] **Test coverage: The changes of this PR are covered by test.** Unit test were added with mocks when depending on a backend/device.
- [x] **No impact: The changes of this PR have ZERO impact on the userland.** Meaning, we can use these changes without modifying LLD/LLM at all. It will be a "noop" and the maintainers will be able to bump it without changing anything.

### CODE REVIEW

File: src/families/bitcoin/js-synchronisation.ts [hunk 1]
```diff
@@ -323,17 +323,18 @@ const getAccountShape: GetAccountShape = async (info) => {
...
+
+  const changeAddresses: Set<string> = new Set<string>();
   const changeAddressesWithInfo =
     await walletAccount.xpub.storage.getUniquesAddresses({
       account: 1,
     });
-  const changeAddresses = changeAddressesWithInfo
-    ? changeAddressesWithInfo.map((a) => a.address)
-    : [];
+  changeAddressesWithInfo.map((a) => changeAddresses.add(a.address));
```
Comment (hzheng-ledger) [inline src/families/bitcoin/js-synchronisation.ts:336]: maybe
const changeAddresses = new Set( changeAddressesWithInfo.map(  a => a.address));
would be more concise.

File: src/families/bitcoin/js-synchronisation.ts [hunk 2]
```diff
@@ -323,17 +323,18 @@ const getAccountShape: GetAccountShape = async (info) => {
   const { txs: transactions } = await wallet.getAccountTransactions(
     walletAccount
   );
+
+  const accountAddresses: Set<string> = new Set<string>();
   const accountAddressesWithInfo = await walletAccount.xpub.getXpubAddresses();
-  const accountAddresses = accountAddressesWithInfo
-    ? accountAddressesWithInfo.map((a) => a.address)
-    : [];
+  accountAddressesWithInfo.map((a) => accountAddresses.add(a.address));
```
Comment (hzheng-ledger) [inline src/families/bitcoin/js-synchronisation.ts:329]: maybe
const accountAddresses = new Set( accountAddresses.map(  a => a.address));
would be more concise.

**Review by hzheng-ledger — APPROVED** [review]
looks good to me

**Comment by haammar-ledger** [comment]
Replaced with https://github.com/LedgerHQ/ledger-live-common/pull/1599

**Pull request closed without merging**